src/drivers/ins/microstrain/MicroStrain.cpp

```diff
@@ -0,0 +1,623 @@
+	_sensor_baro.device_id = dev_id;
+	_sensor_baro.pressure = 0;
+	_sensor_baro.temperature = 0;
+	_sensor_baro.error_count = 0;
+}
+
+MicroStrain::~MicroStrain()
+{
+	if (device_uart.is_open()) {
+		device_uart.uart_close();
+	}
+
+	PX4_INFO("Destructor");
+	_sensor_baro_pub.unadvertise();
+
+	perf_free(_loop_perf);
+	perf_free(_loop_interval_perf);
+}
+
+bool mip_interface_user_recv_from_device(mip_interface *device, uint8_t *buffer, size_t max_length,
+		timeout_type wait_time,
+		size_t *out_length, timestamp_type *timestamp_out)
+{
+	(void)device;
+
+	*timestamp_out = hrt_absolute_time();
+
+	int res = device_uart.uart_read(buffer, max_length);
+
+	if (res == -1 && errno != EAGAIN) {
+		PX4_DEBUG("RX 1 %d(%d)", res, max_length);
+		*out_length = 0;
+		return false;
+	}
+
+	if (res >= 0) {
+		*out_length = res;
+		PX4_DEBUG("RX 2 %d(%d)", *out_length, max_length);
+	}
+
+	PX4_DEBUG("RX 3 %d(%d)", *out_length, max_length);
+
+	return true;
+}
+
+bool mip_interface_user_send_to_device(mip_interface *device, const uint8_t *data, size_t length)
+{
+
+	int res = device_uart.uart_write(const_cast<uint8_t *>(data), length);
+
+	if (res >= 0) {
+		return true;
+	}
+
+	return false;
+
+}
+
+int MicroStrain::connect_at_baud(int32_t baud)
+{
+	if (device_uart.is_open()) {
+		if (device_uart.uart_set_baud(baud) == PX4_ERROR) {
+			PX4_INFO(" - Failed to set UART %" PRIu32 " baud", baud);
+		}
+
+	} else if (device_uart.uart_open(_port, baud) == PX4_ERROR) {
+		PX4_INFO(" - Failed to open UART");
+		PX4_ERR("ERROR: Could not open device port!");
+		return PX4_ERROR;
+	}
+
+	PX4_INFO("Serial Port %s with baud of %" PRIu32 " baud", (device_uart.is_open() ? "CONNECTED" : "NOT CONNECTED"), baud);
+
+	// Re-init the interface with the correct timeouts
+	mip_interface_init(&device, parse_buffer, sizeof(parse_buffer), mip_timeout_from_baudrate(baud) * 1_ms, 250_ms,
+			   &mip_interface_user_send_to_device, &mip_interface_user_recv_from_device, &mip_interface_default_update, NULL);
+
+	PX4_INFO("mip_base_ping");
+
+	if (mip_base_ping(&device) != MIP_ACK_OK) {
+		PX4_INFO(" - Failed to Ping 1");
+		usleep(200_ms);
+
+		if (mip_base_ping(&device) != MIP_ACK_OK) {
+			PX4_INFO(" - Failed to Ping 2");
+			return PX4_ERROR;
+		}
+	}
+
+	PX4_INFO("Successfully opened and pinged");
+	return PX4_OK;
+}
+
+mip_cmd_result MicroStrain::get_supported_descriptors()
+{
+	const size_t descriptors_max_size = sizeof(_supported_descriptors) / sizeof(_supported_descriptors[0]);
+	uint8_t descriptors_count, extended_descriptors_count;
+
+	mip_cmd_result res = mip_base_get_device_descriptors(&device, _supported_descriptors, descriptors_max_size,
+			     &descriptors_count);
+	mip_cmd_result res_extended = mip_base_get_extended_descriptors(&device, &(_supported_descriptors[descriptors_count]),
+				      descriptors_max_size - descriptors_count, &extended_descriptors_count);
+
+	if (res != MIP_ACK_OK || res_extended != MIP_ACK_OK) {
```

Comment:
Not sure if we plan to support older devices, but not all devices support the extended descriptors command. [Here](https://github.com/LORD-MicroStrain/microstrain_inertial_driver_common/blob/main/src/utils/mip/ros_mip_device_main.cpp#L178) is how I get the supported descriptors in ROS